trickle/trickle_server.go

```diff
@@ -0,0 +1,423 @@
+package trickle
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+)
+
+// TODO sweep idle streams connections
+
+type StreamManager struct {
+	mutex   sync.RWMutex
+	streams map[string]*Stream
+}
+
+type Stream struct {
+	mutex       sync.RWMutex
+	segments    []*Segment
+	latestWrite int
+}
+
+type Segment struct {
+	idx    int
+	mutex  *sync.Mutex
+	cond   *sync.Cond
+	buffer *bytes.Buffer
+	closed bool
+}
+
+type SegmentSubscriber struct {
+	segment *Segment
+	readPos int
+}
+
+const maxSegmentsPerStream = 5
+
+const BaseServerPath = "/ai/live-video/"
+
+var FirstByteTimeout = errors.New("pending read timeout")
+
+func ConfigureServerWithMux(mux *http.ServeMux) {
+	/* TODO we probably want to configure the below
+	srv := &http.Server{
+		// say max segment size is 20 secs
+		// we can allow 2 * 20 secs given preconnects
+		ReadTimeout:  40 * time.Second,
+		WriteTimeout: 45 * time.Second,
+	}
+	*/
+
+	streamManager := &StreamManager{
+		streams: make(map[string]*Stream),
+	}
+	mux.HandleFunc("GET "+BaseServerPath+"{streamName}/{idx}", streamManager.handleGet)
+	mux.HandleFunc("POST "+BaseServerPath+"{streamName}/{idx}", streamManager.handlePost)
+	mux.HandleFunc("DELETE "+BaseServerPath+"{streamName}", streamManager.handleDelete)
+}
+
+func (sm *StreamManager) getStream(streamName string) (*Stream, bool) {
+	sm.mutex.RLock()
+	defer sm.mutex.RUnlock()
+	stream, exists := sm.streams[streamName]
+	return stream, exists
+}
+
+func (sm *StreamManager) getOrCreateStream(streamName string) *Stream {
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+
+	stream, exists := sm.streams[streamName]
+	if !exists {
+		stream = &Stream{
+			segments: make([]*Segment, 5),
```

Comment:
```suggestion
			segments: make([]*Segment, maxSegmentsPerStream),
```

Shouldn't you use `maxSegmentsPerStream` instead of `5` here?

Reply:
Yep, fixed in cccdde875b2702c502e5a379accf61dcc0e91488